### PR TITLE
VSTS-349 Add retry mechanism during polling

### PR DESCRIPTION
**Pull request by 7PH**
Here is the output in the terminal when the request fails (once):

```text
2024-01-16T12:57:37.7455381Z ##[debug][SQ] API GET: '/api/ce/task' with full URL "https://xxx.eu.ngrok.io/api/ce/task?id=AY0SWUK2VqVwuP_SOuqo" and query "{"id":"AY0SWUK2VqVwuP_SOuqo"}"
2024-01-16T12:57:37.8555253Z ##[debug][SQ] Task status:IN_PROGRESS
2024-01-16T12:57:38.8580785Z ##[debug][SQ] API GET: '/api/ce/task' with full URL "https://xxx.eu.ngrok.io/api/ce/task?id=AY0SWUK2VqVwuP_SOuqo" and query "{"id":"AY0SWUK2VqVwuP_SOuqo"}"
2024-01-16T12:57:38.9892444Z ##[debug][SQ] Task status:IN_PROGRESS
2024-01-16T12:57:39.9909318Z ##[debug][SQ] API GET: '/api/ce/task' with full URL "https://xxx.eu.ngrok.io/api/ce/task?id=AY0SWUK2VqVwuP_SOuqo" and query "{"id":"AY0SWUK2VqVwuP_SOuqo"}"
2024-01-16T12:57:40.1174809Z ##[debug][SQ] Task status:IN_PROGRESS
2024-01-16T12:57:41.1179568Z ##[debug][SQ] API GET: '/api/ce/task' with full URL "https://xxx.eu.ngrok.io/api/ce/task?id=AY0SWUK2VqVwuP_SOuqo" and query "{"id":"AY0SWUK2VqVwuP_SOuqo"}"
2024-01-16T12:57:41.2377917Z ##[debug][SQ] API GET '/api/ce/task' failed, error is invalid json response body at https://xxx.eu.ngrok.io/api/ce/task?id=AY0SWUK2VqVwuP_SOuqo reason: Unexpected token < in JSON at position 0
2024-01-16T12:57:41.2379089Z ##[debug][SQ] API GET '/api/ce/task' failed (attempt 1/3)
2024-01-16T12:57:43.2407634Z ##[debug][SQ] API GET: '/api/ce/task' with full URL "https://xxx.eu.ngrok.io/api/ce/task?id=AY0SWUK2VqVwuP_SOuqo" and query "{"id":"AY0SWUK2VqVwuP_SOuqo"}"
2024-01-16T12:57:43.3708339Z ##[debug][SQ] Task status:SUCCESS
2024-01-16T12:57:43.3710283Z ##[debug][SQ] Task complete:  [..]
```

If it fails consistently:

```text
##[debug][SQ] API GET: '/api/ce/task' with full URL "[https://xxx.eu.ngrok.io/api/ce/task?id=AY0SXqBEVqVwuP_SOuqx"](https://xxx.eu.ngrok.io/api/ce/task?id=AY0SXqBEVqVwuP_SOuqx%22) and query "{"id":"AY0SXqBEVqVwuP_SOuqx"}"
##[debug][SQ] Task status:IN_PROGRESS
##[debug][SQ] API GET: '/api/ce/task' with full URL "[https://xxx.eu.ngrok.io/api/ce/task?id=AY0SXqBEVqVwuP_SOuqx"](https://xxx.eu.ngrok.io/api/ce/task?id=AY0SXqBEVqVwuP_SOuqx%22) and query "{"id":"AY0SXqBEVqVwuP_SOuqx"}"
##[debug][SQ] API GET '/api/ce/task' failed, error is invalid json response body at https://xxx.eu.ngrok.io/api/ce/task?id=AY0SXqBEVqVwuP_SOuqx reason: Unexpected token < in JSON at position 0
##[debug][SQ] API GET '/api/ce/task' failed (attempt 1/3)
##[debug][SQ] API GET: '/api/ce/task' with full URL "[https://xxx.eu.ngrok.io/api/ce/task?id=AY0SXqBEVqVwuP_SOuqx"](https://xxx.eu.ngrok.io/api/ce/task?id=AY0SXqBEVqVwuP_SOuqx%22) and query "{"id":"AY0SXqBEVqVwuP_SOuqx"}"
##[debug][SQ] API GET '/api/ce/task' failed, error is invalid json response body at https://xxx.eu.ngrok.io/api/ce/task?id=AY0SXqBEVqVwuP_SOuqx reason: Unexpected token < in JSON at position 0
##[debug][SQ] API GET '/api/ce/task' failed (attempt 2/3)
##[debug][SQ] API GET: '/api/ce/task' with full URL "[https://xxx.eu.ngrok.io/api/ce/task?id=AY0SXqBEVqVwuP_SOuqx"](https://xxx.eu.ngrok.io/api/ce/task?id=AY0SXqBEVqVwuP_SOuqx%22) and query "{"id":"AY0SXqBEVqVwuP_SOuqx"}"
##[debug][SQ] API GET '/api/ce/task' failed, error is invalid json response body at https://xxx.eu.ngrok.io/api/ce/task?id=AY0SXqBEVqVwuP_SOuqx reason: Unexpected token < in JSON at position 0
##[debug][SQ] API GET '/api/ce/task' failed (attempt 3/3)
##[error][SQ] API GET '/api/ce/task' failed, max attempts reached
##[debug]Processed: ##vso[task.issue type=error;][SQ] API GET '/api/ce/task' failed, max attempts reached
##[debug][SQ] Publish task error: [SQ] Could not fetch task for ID 'AY0SXqBEVqVwuP_SOuqx'
##[debug]task result: Failed
##[error][SQ] Could not fetch task for ID 'AY0SXqBEVqVwuP_SOuqx'
```

### CODE REVIEW

File: commonv5/ts/helpers/__tests__/api-test.ts [hunk 1]
```diff
@@ -43,8 +43,44 @@ jest.mock("../../helpers/request", () => ({
...
-  (get as jest.Mock<any>).mockClear();
+describe("fetchWithRetry", () => {
+  beforeEach(() => {
+    (get as jest.Mock<any>).mockClear();
+  });
+
+  it(
+    "should not fail after up to 2 attempts",
+    async () => {
+      for (let i = 0; i < 2 - 1; i++) {
```
Comment (stanislavhh) [inline commonv5/ts/helpers/__tests__/api-test.ts:54]: NP: I assume you wanted to run 2 failures here

```suggestion
      for (let i = 0; i < 2; i++) {
```

File: commonv5/ts/helpers/__tests__/api-test.ts [hunk 2]
```diff
@@ -43,8 +43,44 @@ jest.mock("../../helpers/request", () => ({
...
+    "should fail after 3 failing requests",
+    async () => {
+      for (let i = 0; i < 3; i++) {
+        jest.mocked(get).mockRejectedValueOnce(new Error("foo"));
+      }
+
+      expect.assertions(2);
+      try {
+        const response = await fetchWithRetry(MOCKED_ENDPOINT, "/api", false);
+        console.log(response);
```
Comment (stanislavhh) [inline commonv5/ts/helpers/__tests__/api-test.ts:76]: NP: lets drop it/

**Review by stanislavhh — APPROVED** [review]
Nice work and big kudos for the coverage!

File: commonv5/ts/helpers/__tests__/api-test.ts (hunk 1 above)
Comment (7PH) [reply commonv5/ts/helpers/__tests__/api-test.ts:54]: Thanks for your due diligence 🙏

**Pull request merged**